autoray/autoray.py

```diff
@@ -419,28 +409,75 @@ def complex_add_re_im(re, im):
     return re + 1j * im
 
 
+# ----------------------------- Custom dispatchers -------------------------- #
+
+
+def register_dispatch(fun, dispatcher):
+    _DISPATCHERS[fun] = dispatcher
+
+
+def default_dispatcher(*args, **kwargs):
+    return args[0]
+
+
+# lookup of custom dispatcher methods, for cases when backend cannot be inferred
+#     accurately from first argument.
+_DISPATCHERS = defaultdict(lambda: default_dispatcher)
+
+
+# Numpy array joining operations always take a sequence of arrays as input.
+#     The backend is inferred from the first element of this sequence. Numpy
+#     always assumes that the argument is iterbale, so calling args[0][0]
+#     should be safe.
+def join_array_dispatcher(*args, **kwargs):
+    # NOTE: To be super safe, we could do a try / except, or
+    #       hasattr(args[0], '__getitem__'). hasattr is then probably better,
+    #       since there would be multiple types of errors to catch
+    return args[0][0]

```

Comment:
One solution here that might? solve the two problems you mention is:
```python
def join_array_dispatcher(arg0, *args, **kwargs):
    if infer_backend(arg0)== 'builtins':
        return join_array_dispatcher(arg0[0])
    return arg0
```
though you will still get an error with an e.g. empty tuple.

Reply:
This is also not perfect. It requires additional lookups, whereas the situation where args[0][0] is not defined is rare and not intended usage (at least within numpy API). Furthermore this will fail in the case that the sequences is not of `builtins` type, for example if you use a numpy object array containing e.g. torch tensors.  
Granted that's not a common use-case, but in those cases it will silently convert the tensors to the wrong type, and I think it's better to just throw an error, because in any case you should not call these functions with an empty tuple or one-dimensional array.

I think we should go for something like

```python
def join_array_dispatcher(*args, **kwargs):
    try:
        return args[0][0]
    except (TypeError, IndexError):
       raise ValueError("array joining function should not be called with empty argument")
```

This way the user gets an informative error if called with wrong argument. Overhead of error handling should be less than the additional lookup you proposed, although I haven't timed it.  
This still will silently convert to numpy backend if called with a one-dimensional `sparse` array. I think that's however fine.

Reply:
Sure thing, I'm happy with this logic also.